### PR TITLE
Convert to Swift 3

### PR DESCRIPTION
**Pull request by derpoliuk**

### CODE REVIEW

**Comment by derpoliuk** [comment]
Ready for review. CC @salutis @mikemee

**Review by mikemee — APPROVED** [review]
@derpoliuk Nice work Staas.

I'm approving, but as the first Swift 3 conversion, I'd like @salutis to give his blessing too. Thx!

**Comment by mikemee** [comment]
Not sure why travis wants to build again, again, but ... at least one worked.

**Comment by rudolf-adamkovic** [comment]
In review.

File: ReceiptKit/Classes/ReceiptRequest.swift [hunk 1]
```diff
@@ -102,18 +102,18 @@ private extension ReceiptRequest {
 
 private extension ReceiptRequest {
 
-    var URL: NSURL {
-        guard let URL = NSURL(string: URLString) else {
+    var URL: Foundation.URL {
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/ReceiptRequest.swift]: No need to disambiguate. `Foundation.URL` can be `URL`.

File: ReceiptKit/Classes/ReceiptRequest.swift [hunk 2]
```diff
@@ -102,18 +102,18 @@ private extension ReceiptRequest {
...
+        guard let URL = Foundation.URL(string: URLString) else {
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/ReceiptRequest.swift]: No need to disambiguate. `Foundation.URL` can be `URL`.

File: ReceiptKit/Classes/ReceiptRequest.swift [hunk 3]
```diff
@@ -73,24 +73,24 @@ private extension ReceiptRequest {
 
 private extension ReceiptRequest {
 
-    var request: NSURLRequest {
-        let request = NSMutableURLRequest(URL: URL)
-        request.HTTPMethod = "POST"
-        request.HTTPBody = requestData
-        return request
+    var request: URLRequest {
+        let request = NSMutableURLRequest(url: URL)
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/ReceiptRequest.swift]: Needs to be converted to `URLRequest`:

Old:

``` swift
let request = NSMutableURLRequest(url: URL)
```

New:

``` swift
var request = URLRequest(url: URL)
```

Once done, we want to also remove now unnecessary cast `as URLRequest` below.

File: ReceiptKit/Classes/ReceiptRequest.swift [hunk 4]
```diff
@@ -51,20 +51,20 @@ extension ReceiptRequest {
...
 private extension ReceiptRequest {
 
-    func newDataTask() -> NSURLSessionDataTask {
-        return session.dataTaskWithRequest(request) { [delegate, queue] data, response, error in
+    func newDataTask() -> URLSessionDataTask {
+        return session.dataTask(with: request, completionHandler: { [delegate, queue] data, response, error in
             var JSONObject: AnyObject? {
                 guard let data = data else { return nil }
-                return try? NSJSONSerialization.JSONObjectWithData(data, options: [])
+                return try! JSONSerialization.jsonObject(with: data, options: []) as AnyObject?
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/ReceiptRequest.swift]: Introduces unnecessary crashes.

Old:

``` swift
return try! JSONSerialization.jsonObject(with: data, options: []) as AnyObject?
```

New:

``` swift
return try? JSONSerialization.jsonObject(with: data, options: []) as AnyObject
```

File: ReceiptKit/Classes/ReceiptRequest.swift [hunk 5]
```diff
@@ -51,20 +51,20 @@ extension ReceiptRequest {
 
 private extension ReceiptRequest {
 
-    func newDataTask() -> NSURLSessionDataTask {
-        return session.dataTaskWithRequest(request) { [delegate, queue] data, response, error in
+    func newDataTask() -> URLSessionDataTask {
+        return session.dataTask(with: request, completionHandler: { [delegate, queue] data, response, error in
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/ReceiptRequest.swift]: This reads better as a trailing closure.

File: ReceiptKit/Classes/ReceiptRequest.swift [hunk 6]
```diff
@@ -6,23 +6,23 @@ import Foundation
 
 protocol ReceiptRequestDelegate: class {
 
-    func receiptRequest(receiptRequest: ReceiptRequest, didRetrieveReceipt receipt: Receipt?)
+    func receiptRequest(_ receiptRequest: ReceiptRequest, didRetrieveReceipt receipt: Receipt?)
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/ReceiptRequest.swift]: Unnecessary repetition of type information.

Current:

``` swift
func receiptRequest(_ receiptRequest: ReceiptRequest, didRetrieveReceipt receipt: Receipt?)
```

Suggested:

``` swift
func receiptRequest(_ receiptRequest: ReceiptRequest, didRetrieve receipt: Receipt?)
```

File: ReceiptKit/Classes/ReceiptDataFormatter.swift [hunk 7]
```diff
@@ -8,11 +8,11 @@ final class ReceiptDateFormatter {}
 
 extension ReceiptDateFormatter {
 
-    static func dateFromString(string: String) -> NSDate? {
-        let dateFormatter = NSDateFormatter()
-        dateFormatter.locale = NSLocale(localeIdentifier: ReceiptDateFormatter.locale)
+    static func dateFromString(_ string: String) -> Date? {
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/ReceiptDataFormatter.swift]: To match `DateFormatter`, this should be `date(from string: String)`.

File: ReceiptKit/Classes/Receipt.swift [hunk 8]
```diff
@@ -38,21 +38,21 @@ public extension Receipt {
...
+    fileprivate static let environmentKey = "environment"
 
 }
 
 // MARK: Parse purchase date
 
 public extension Receipt {
 
-    var purchaseDate: NSDate? {
+    var purchaseDate: Date? {
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/Receipt.swift]: Unnecessary repetition of type information. Better as `purchasedAt` in Swift 3.

File: ReceiptKit/Classes/CompoundReceiptRequest.swift [hunk 9]
```diff
@@ -6,18 +6,18 @@ import Foundation
 
 protocol CompoundReceiptRequestDelegate: class {
 
-    func compoundReceiptRequest(compoundReceiptRequest: CompoundReceiptRequest, didRetrieveReceipt receipt: Receipt?)
+    func compoundReceiptRequest(_ compoundReceiptRequest: CompoundReceiptRequest, didRetrieveReceipt receipt: Receipt?)
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/CompoundReceiptRequest.swift]: Unnecessary repetition of type information.

Current:

``` swift
func compoundReceiptRequest(_ compoundReceiptRequest: CompoundReceiptRequest, didRetrieveReceipt receipt: Receipt?)
```

Suggested:

``` swift
func compoundReceiptRequest(_ compoundReceiptRequest: CompoundReceiptRequest, didRetrieve receipt: Receipt?)
```

File: ReceiptKit/Classes/BundleReceiptRequest.swift [hunk 10]
```diff
@@ -50,12 +50,12 @@ extension BundleReceiptRequest: CompoundReceiptRequestDelegate {
 
 private extension BundleReceiptRequest {
 
-    var receiptData: NSData? {
+    var receiptData: Data? {
         guard let receiptURL = receiptURL else { return nil }
-        return NSData(contentsOfURL: receiptURL)
+        return (try? Data(contentsOf: receiptURL))
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/BundleReceiptRequest.swift]: Unnecessary parentheses around `try?` expression.

File: ReceiptKit/Classes/BundleReceiptRequest.swift [hunk 11]
```diff
@@ -6,18 +6,18 @@ import StoreKit
 
 public protocol BundleReceiptRequestDelegate: class {
 
-    func bundleReceiptRequest(bundleReceiptRequest: BundleReceiptRequest, didRetrieveReceipt receipt: Receipt?)
+    func bundleReceiptRequest(_ bundleReceiptRequest: BundleReceiptRequest, didRetrieveReceipt receipt: Receipt?)
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/BundleReceiptRequest.swift]: Unnecessary repetition of type information.

Current:

``` swift
func bundleReceiptRequest(_ bundleReceiptRequest: BundleReceiptRequest, didRetrieveReceipt receipt: Receipt?)
```

Suggested:

``` swift
func bundleReceiptRequest(_ bundleReceiptRequest: BundleReceiptRequest, didRetrieve receipt: Receipt?)
```

File: ReceiptKit/Classes/BundleReceiptRequest.swift [hunk 12]
```diff
@@ -6,18 +6,18 @@ import StoreKit
...
 // MARK: Main
 
 public final class BundleReceiptRequest: NSObject {
 
-    public let bundle: NSBundle
+    public let bundle: Bundle
     public weak var delegate: BundleReceiptRequestDelegate?
 
-    public init(bundle: NSBundle = NSBundle.mainBundle(), delegate: BundleReceiptRequestDelegate? = nil) {
+    public init(bundle: Bundle = Bundle.main, delegate: BundleReceiptRequestDelegate? = nil) {
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/BundleReceiptRequest.swift]: `Bundle.main` can be just `.main`.

File: ReceiptKit/Classes/CompoundReceiptRequest.swift [hunk 13]
```diff
@@ -58,32 +58,32 @@ extension CompoundReceiptRequest {
...
+    func receiptRequest(_ receiptRequest: ReceiptRequest, didRetrieveReceipt receipt: Receipt?) {
         didFinishParsingReceipt(receipt, type: receiptRequest.receiptType)
     }
 
 }
 
 private extension CompoundReceiptRequest {
 
-    func didFinishParsingReceipt(receipt: Receipt?, type: ReceiptType) {
+    func didFinishParsingReceipt(_ receipt: Receipt?, type: ReceiptType) {
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/CompoundReceiptRequest.swift]: Needs update.

Current:

``` swift
func didFinishParsingReceipt(_ receipt: Receipt?, type: ReceiptType)
```

Suggested:

``` swift
func didFinishParsing(receipt: Receipt?, type: ReceiptType)
```

File: ReceiptKit/Classes/CompoundReceiptRequest.swift [hunk 14]
```diff
@@ -58,32 +58,32 @@ extension CompoundReceiptRequest {
...
+        case .production:
             didFinishParsingProductionReceipt(receipt)
-        case .Sandbox:
+        case .sandbox:
             didFinishParsingSandboxReceipt(receipt)
         }
     }
 
-    func didFinishParsingProductionReceipt(receipt: Receipt?) {
+    func didFinishParsingProductionReceipt(_ receipt: Receipt?) {
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/CompoundReceiptRequest.swift]: Needs update.

Current:

``` swift
func didFinishParsingProductionReceipt(_ receipt: Receipt?)
```

Suggested:

``` swift
func didFinishParsing(productionReceipt: Receipt?)
```

File: ReceiptKit/Classes/CompoundReceiptRequest.swift [hunk 15]
```diff
@@ -58,32 +58,32 @@ extension CompoundReceiptRequest {
...
         if let receipt = receipt {
             delegate?.compoundReceiptRequest(self, didRetrieveReceipt: receipt)
         } else {
-            beginParsingType(.Sandbox)
+            beginParsingType(.sandbox)
         }
     }
 
-    func didFinishParsingSandboxReceipt(receipt: Receipt?) {
+    func didFinishParsingSandboxReceipt(_ receipt: Receipt?) {
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/CompoundReceiptRequest.swift]: Needs update.

Current:

``` swift
func didFinishParsingSandboxReceipt(_ receipt: Receipt?)
```

Suggested:

``` swift
func didFinishParsing(sandboxReceipt: Receipt?)
```

File: ReceiptKit/Classes/CompoundReceiptRequest.swift [hunk 16]
```diff
@@ -46,7 +46,7 @@ private extension CompoundReceiptRequest {
 
 extension CompoundReceiptRequest {
 
-    func instantiateReceiptRequestWithReceiptType(receiptType: ReceiptType) -> ReceiptRequest {
+    func instantiateReceiptRequestWithReceiptType(_ receiptType: ReceiptType) -> ReceiptRequest {
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/CompoundReceiptRequest.swift]: This method doesn't play well with Swift 3. I would suggest to inline it.

File: ReceiptKit/Classes/CompoundReceiptRequest.swift [hunk 17]
```diff
@@ -28,14 +28,14 @@ final class CompoundReceiptRequest {
...
-        beginParsingType(.Production)
+        beginParsingType(.production)
     }
 
 }
 
 private extension CompoundReceiptRequest {
 
-    func beginParsingType(receiptType: ReceiptType) {
+    func beginParsingType(_ receiptType: ReceiptType) {
```
Comment (rudolf-adamkovic) [inline ReceiptKit/Classes/CompoundReceiptRequest.swift]: Needs update.

Current:

``` swift
func beginParsingType(_ receiptType: ReceiptType)
```

Suggested:

``` swift
func beginParsing(receiptType: ReceiptType)
```

**Review by rudolf-adamkovic — CHANGES_REQUESTED** [review]

**Comment by derpoliuk** [comment]
@salutis thank you for feedback. These are automatic changes. I'll update code.

**Comment by derpoliuk** [comment]
@salutis I addressed your comments. Ready for review.

**Review by rudolf-adamkovic — CHANGES_REQUESTED** [review]
@derpoliuk LGTM except some of your commit messages:
- `Refactor CompoundReceiptRequest` — too vague, used 4x
- `Inline CompoundReceiptRequest.instantiateReceiptRequestWithReceiptType(_:)` — too long
- `Refactor URL and URLRequest usage in ReceipRequest` — contains typo

**Comment by derpoliuk** [comment]
@salutis thank you for review and feedback.

> `Refactor CompoundReceiptRequest` — too vague, used 4x

There were commit descriptions with detailed info. But I squashed "Refactor CompoundReceiptRequest [...]`into`Remove unnecessary type repetition in CompoundReceiptRequest`.

> `Inline CompoundReceiptRequest.instantiateReceiptRequestWithReceiptType(_:)`

Renamed to `Inline ReceiptRequest instantiation`

> `Refactor URL and URLRequest usage in ReceipRequest` — contains typo

Fixed.

**Comment by rudolf-adamkovic** [comment]
In review.

**Review by rudolf-adamkovic — CHANGES_REQUESTED** [review]
@derpoliuk One last typo to fix:

`Conver to Swift 3` (7ec020ef192be3691f9e6009c128a1458fa0fae2)

**Comment by derpoliuk** [comment]
@salutis thank you. I renamed commit.

**Review by rudolf-adamkovic — APPROVED** [review]

**Comment by rudolf-adamkovic** [comment]
@derpoliuk Doesn't compile on CI. I guess you will need to update the `xcode7.3` line in `.travis.yml`.

**Comment by derpoliuk** [comment]
@salutis thanks!

**Pull request merged**